CMakeLists.txt

```diff
@@ -386,29 +386,74 @@ function(check_working_std_regex result_var_name)
   else()
-    # libstdc++ 4.8 has an incomplete std::regex implementation, and crashes
-    # on many regexes.
-    # libstdc++ 4.9 works.
-    set(std_regex_test_source
-"
-#include <regex>
-const std::regex broken_regex{
-  \"([a]+)\",
-  std::regex::ECMAScript | std::regex::nosubs};
-
-int main() {}
-")
-
-    check_cxx_source_runs("${std_regex_test_source}" "${result_var_name}_TEST")
-    if ("${${result_var_name}_TEST}")
-      set("${result_var_name}" TRUE PARENT_SCOPE)
-    else()
+    if(CMAKE_CROSSCOMPILING)
+      # Can't run C source when cross-compiling; assume false until we have a static check.
       set("${result_var_name}" FALSE PARENT_SCOPE)
+    else()
+      # libstdc++ 4.8 has an incomplete std::regex implementation, and crashes
+      # on many regexes.
+      # libstdc++ 4.9 works.
+      set(std_regex_test_source
+  "
+  #include <regex>
+  const std::regex broken_regex{
+    \"([a]+)\",
+    std::regex::ECMAScript | std::regex::nosubs};
+
+  int main() {}
+  ")
+
+      check_cxx_source_runs("${std_regex_test_source}" "${result_var_name}_TEST")
+      if ("${${result_var_name}_TEST}")
+        set("${result_var_name}" TRUE PARENT_SCOPE)
+      else()
+        set("${result_var_name}" FALSE PARENT_SCOPE)
+      endif()
     endif()
   endif()
 endfunction()
 check_working_std_regex(SWIFT_HAVE_WORKING_STD_REGEX)
 
+# If SWIFT_HOST_VARIANT_SDK not given, try to detect from the CMAKE_SYSTEM_NAME
+if("${SWIFT_HOST_VARIANT_SDK}" STREQUAL "")
+  if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
+    set(SWIFT_HOST_VARIANT_SDK "LINUX")
+  elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "FreeBSD")
+    set(SWIFT_HOST_VARIANT_SDK "FREEBSD")
+  elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "CYGWIN")
+    set(SWIFT_HOST_VARIANT_SDK "CYGWIN")
+  elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
+    set(SWIFT_HOST_VARIANT_SDK "OSX")
+  else()
+    message(FATAL_ERROR "Unable to detect SDK for host system: ${CMAKE_SYSTEM_NAME}")
+  endif()
+endif()
+
+# If SWIFT_HOST_VARIANT_ARCH not given, try to detect from the CMAKE_SYSTEM_PROCESSOR
+if("${SWIFT_HOST_VARIANT_ARCH}" STREQUAL "")
+  if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
+    set(SWIFT_HOST_VARIANT_ARCH "x86_64")
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "aarch64")
+    set(SWIFT_HOST_VARIANT_ARCH "aarch64")
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "ppc64")
+    set(SWIFT_HOST_VARIANT_ARCH "powerpc64")
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "ppc64le")
+    set(SWIFT_HOST_VARIANT_ARCH "powerpc64le")
+  # FIXME: Only matches v6l/v7l - by far the most common variants
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv6l")
+    set(SWIFT_HOST_VARIANT_ARCH "armv6")
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv7l")
+    set(SWIFT_HOST_VARIANT_ARCH "armv7")
+  else()
+    message(FATAL_ERROR "Unrecognized architecture on host system: ${CMAKE_SYSTEM_PROCESSOR}")
+  endif()
+endif()
+
+set(SWIFT_HOST_VARIANT_SDK "${SWIFT_HOST_VARIANT_SDK}" CACHE STRING
```

Comment:
Reusing the same variable name looks very confusing.  Could you change the detection logic above to be unconditional, and store the result into a different variable, say, `SWIFT_HOST_VARIANT_SDK_default`?